### PR TITLE
[common,crypto] Mapping key size to hash alg

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

**Review by mykola-kobets-epam — APPROVED** [review]

File: include/aos/common/crypto/mbedtls/driverwrapper.hpp [hunk 1]
```diff
@@ -12,14 +12,28 @@
...
+ */
+struct KeyInfo {
+    /**
+     * Key ID.
+     */
+    psa_key_id_t mKeyID;
+    /**
+     * Key type.
+     */
+    mbedtls_md_type_t mMdType;
```
Comment (al1img) [inline include/aos/common/crypto/mbedtls/driverwrapper.hpp:26]: mMDType;

File: include/aos/common/crypto/mbedtls/driverwrapper.hpp [hunk 2]
```diff
@@ -12,14 +12,28 @@
...
+/**
+ * Key description.
+ */
+struct KeyInfo {
+    /**
+     * Key ID.
+     */
+    psa_key_id_t mKeyID;
+    /**
+     * Key type.
```
Comment (al1img) [inline include/aos/common/crypto/mbedtls/driverwrapper.hpp:24]: Message digest type.

File: src/common/crypto/mbedtls/driverwrapper.cpp [hunk 3]
```diff
@@ -23,6 +23,21 @@
...
 
+struct AlgorithmDescription {
+    /**
+     * PSA algorithm.
+     */
+    psa_algorithm_t mAlg;
+    /**
+     * Message digest type.
+     */
+    mbedtls_md_type_t mMdType;
```
Comment (al1img) [inline src/common/crypto/mbedtls/driverwrapper.cpp:34]: mMDType

File: src/common/crypto/mbedtls/driverwrapper.cpp [hunk 4]
```diff
@@ -213,11 +232,78 @@ static int ExportECPublicKeyToDER(
...
+    }
+
+    if (curveBitlen <= 384) {
+        return {PSA_ALG_SHA_384, MBEDTLS_MD_SHA384, aos::crypto::HashEnum::eSHA384};
+    }
+
+    return {PSA_ALG_SHA_512, MBEDTLS_MD_SHA512, aos::crypto::HashEnum::eSHA512};
+}
+
+aos::RetWithError<AlgorithmDescription> RSAAlgorithmDescription(const aos::crypto::RSAPublicKey& pubKey)
```
Comment (al1img) [inline src/common/crypto/mbedtls/driverwrapper.cpp:273]: `GetRSAAlgFromPubKey`

File: src/common/crypto/mbedtls/driverwrapper.cpp [hunk 5]
```diff
@@ -213,11 +232,78 @@ static int ExportECPublicKeyToDER(
...
+    }
+
+    auto algDesc = RSASHAAlgorithm(mbedtls_mpi_bitlen(&n));
+
+    mbedtls_mpi_free(&n);
+
+    return aos::RetWithError<AlgorithmDescription>(algDesc);
+}
+
+aos::RetWithError<AlgorithmDescription> ECCAlgorithmDescription(const aos::crypto::ECDSAPublicKey& pubKey)
```
Comment (al1img) [inline src/common/crypto/mbedtls/driverwrapper.cpp:292]: ditto

File: src/common/crypto/mbedtls/driverwrapper.cpp [hunk 6]
```diff
@@ -213,11 +232,78 @@ static int ExportECPublicKeyToDER(
...
+    }
+
+    if (modulusBitlen <= 7680) {
+        return {PSA_ALG_SHA_384, MBEDTLS_MD_SHA384, aos::crypto::HashEnum::eSHA384};
+    }
+
+    return {PSA_ALG_SHA_512, MBEDTLS_MD_SHA512, aos::crypto::HashEnum::eSHA512};
+}
+
+AlgorithmDescription ECCSHAAlgorithm(size_t curveBitlen)
```
Comment (al1img) [inline src/common/crypto/mbedtls/driverwrapper.cpp:252]: ditto

File: src/common/crypto/mbedtls/driverwrapper.cpp [hunk 7]
```diff
@@ -213,11 +232,78 @@ static int ExportECPublicKeyToDER(
     return 0;
 }
 
+AlgorithmDescription RSASHAAlgorithm(size_t modulusBitlen)
```
Comment (al1img) [inline src/common/crypto/mbedtls/driverwrapper.cpp:235]: It is unclear why you always need to keep all three values if you can always get it. For example:
```c++
aos::RetWithError<psa_algorithm_t> GetPSAAlg(aos::crypto::HashEnum hash)
{
...
}

aos::RetWithError<mbedtls_md_type_t> GetMDType(aos::crypto::HashEnum hash)
{
...
}

aos::crypto::HashEnum GetRSASHAAlgorithm(size_t modulusBitlen)
{
...
}
```

**Review by mykola-kobets-epam — APPROVED** [review]

File: include/aos/common/crypto/mbedtls/driverwrapper.hpp (hunk 1 above)
Comment (MykolaSolyanko) [reply include/aos/common/crypto/mbedtls/driverwrapper.hpp]: Done

File: include/aos/common/crypto/mbedtls/driverwrapper.hpp (hunk 2 above)
Comment (MykolaSolyanko) [reply include/aos/common/crypto/mbedtls/driverwrapper.hpp]: Done

File: src/common/crypto/mbedtls/driverwrapper.cpp (hunk 3 above)
Comment (MykolaSolyanko) [reply src/common/crypto/mbedtls/driverwrapper.cpp]: Done

File: src/common/crypto/mbedtls/driverwrapper.cpp (hunk 4 above)
Comment (MykolaSolyanko) [reply src/common/crypto/mbedtls/driverwrapper.cpp]: Done

File: src/common/crypto/mbedtls/driverwrapper.cpp (hunk 5 above)
Comment (MykolaSolyanko) [reply src/common/crypto/mbedtls/driverwrapper.cpp]: Done

File: src/common/crypto/mbedtls/driverwrapper.cpp (hunk 6 above)
Comment (MykolaSolyanko) [reply src/common/crypto/mbedtls/driverwrapper.cpp]: Done

File: src/common/crypto/mbedtls/driverwrapper.cpp (hunk 7 above)
Comment (MykolaSolyanko) [reply src/common/crypto/mbedtls/driverwrapper.cpp]: Done

**Review by al1img — APPROVED** [review]

**Pull request merged**